### PR TITLE
Add type error when attempting to transfer value to a non-payable contract

### LINKED ISSUE #1665: If a contract is not payable, `address(C).send()` should not be available.
Hm, perhaps we could change it to:
the `send` member is available on all addresses and all contracts, but if it is used for a contract that does not have a payable fallback, an error is generated.
And `address(c).send()` does _not_ generate an error, because you can send to any address.

we could also still provide the member but create an error inside the TypeChecker.

### PR DESCRIPTION
**Pull request by federicobond**
Fixes #1665

### CODE REVIEW

File: libsolidity/analysis/TypeChecker.cpp [hunk 1]
```diff
@@ -1489,6 +1489,23 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
...
+	if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
+		if (auto contractType = dynamic_cast<ContractType const*>(type(memberAccess->expression()).get()))
+			if (auto callType = dynamic_cast<FunctionType const*>(type(*memberAccess).get()))
+			{
+				auto kind = callType->kind();
+				auto fallbackFunction = contractType->contractDefinition().fallbackFunction();
+
+				if (
+					(kind == FunctionType::Kind::Send || kind == FunctionType::Kind::Transfer) &&
+					!(fallbackFunction && fallbackFunction->isPayable())
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: It is a personal taste, but for me `(!fallbackFunction || !fallbackFunction->isPayable())` seems more readable.

**Review by axic — CHANGES_REQUESTED** [review]
Please:
- add a test for no fallback function present
- a successful test (payable fallback)
- and a changelog entry

File: test/libsolidity/SolidityNameAndTypeResolution.cpp [hunk 2]
```diff
@@ -6132,6 +6132,42 @@ BOOST_AUTO_TEST_CASE(callable_crash)
...
+	char const* text = R"(
+		contract A {
+			function() {}
+		}
+
+		contract B {
+			A a;
+
+			function() {
+				if (!a.send(100)) throw;
```
Comment (axic) [inline test/libsolidity/SolidityNameAndTypeResolution.cpp]: Change this to `require(a.send(100))` as `throw` issues a deprecation warning now (and will be removed).

**Comment by federicobond** [comment]
@axic updated. Thanks for the review!

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Looks good to me, but @chriseth you might want to double check the `TypeChecker` part.

File: Changelog.md [hunk 3]
```diff
@@ -10,6 +10,7 @@ Bugfixes:
 Features:
  * Syntax Checker: Deprecated "throw" in favour of require(), assert() and revert().
  * Type Checker: Warn if a local storage reference variable does not explicitly use the keyword ``storage``.
+ * Type Checker: Disallow value transfers to contracts without a payable fallback function
```
Comment (chriseth) [inline Changelog.md]: Please update the changelog (move it to 0.4.14).

File: libsolidity/analysis/TypeChecker.cpp [hunk 4]
```diff
@@ -1489,6 +1489,23 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
 		}
 	}
 
+	if (auto memberAccess = dynamic_cast<MemberAccess const*>(&_functionCall.expression()))
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Wouldn't it be better to check already at `MemberAccess`?
As it currently is, `a.transfer` would not generate an error, only `a.transfer(x)` would.

File: libsolidity/analysis/TypeChecker.cpp [hunk 5]
```diff
@@ -1489,6 +1489,23 @@ bool TypeChecker::visit(FunctionCall const& _functionCall)
...
+		if (auto contractType = dynamic_cast<ContractType const*>(type(memberAccess->expression()).get()))
+			if (auto callType = dynamic_cast<FunctionType const*>(type(*memberAccess).get()))
+			{
+				auto kind = callType->kind();
+				auto fallbackFunction = contractType->contractDefinition().fallbackFunction();
+
+				if (
+					(kind == FunctionType::Kind::Send || kind == FunctionType::Kind::Transfer) &&
+					(!fallbackFunction || !fallbackFunction->isPayable())
```
Comment (chriseth) [inline libsolidity/analysis/TypeChecker.cpp]: Can you turn this into a function on the contract type? Something like `isPayable` (on the contract, not the fallback function)?

File: libsolidity/analysis/TypeChecker.cpp (hunk 4 above)
Comment (federicobond) [reply libsolidity/analysis/TypeChecker.cpp]: Makes sense, I started thinking about putting it in the function call at first and never questioned that assumption.

File: libsolidity/analysis/TypeChecker.cpp (hunk 5 above)
Comment (federicobond) [reply libsolidity/analysis/TypeChecker.cpp]: Sure.

File: libsolidity/analysis/TypeChecker.cpp [hunk 6]
```diff
@@ -1622,6 +1622,24 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+				(kind == FunctionType::Kind::Send || kind == FunctionType::Kind::Transfer) &&
+				!contractType->isPayable()
+			)
+				m_errorReporter.typeError(
+					_memberAccess.location(),
+					"Value transfer to a contract with a non-payable fallback function."
+				);
+		}
+	}
 	else if (TypeType const* typeType = dynamic_cast<decltype(typeType)>(exprType.get()))
```
Comment (federicobond) [inline libsolidity/analysis/TypeChecker.cpp]: Double check this else here. I fear there might be some overlap with the code I just added, but I couldn't reproduce it.

Comment (chriseth) [reply libsolidity/analysis/TypeChecker.cpp]: It is probably better to do this at the end of this function. Perhaps even in the `PostTypeChecker`.

File: libsolidity/ast/Types.h [hunk 7]
```diff
@@ -673,6 +673,7 @@ class ContractType: public Type
 	}
 
 	bool isSuper() const { return m_super; }
+	bool isPayable() const;
```
Comment (chriseth) [inline libsolidity/ast/Types.h]: Please add a comment about this returning true if and only if there is a payable fallback function.

**Comment by federicobond** [comment]
@chriseth updated.

File: libsolidity/analysis/TypeChecker.cpp [hunk 8]
```diff
@@ -1628,6 +1628,25 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
...
+			auto contractType = dynamic_cast<ContractType const*>(exprType.get());
+			solAssert(!!contractType, "Should be contract type.");
+
+			if (
+				(kind == FunctionType::Kind::Send || kind == FunctionType::Kind::Transfer) &&
+				!contractType->isPayable()
+			)
+				m_errorReporter.typeError(
+					_memberAccess.location(),
+					"Value transfer to a contract with a non-payable fallback function."
```
Comment (axic) [inline libsolidity/analysis/TypeChecker.cpp]: Perhaps `without a payable fallback function.` is more helpful at showing what needs to be done.

**Comment by axic** [comment]
> Expected message "Value transfer to a contract with a non-payable fallback function." but foundValue transfer to a contract without a payable fallback function.

ie. need to change the expected message in tests.

Do you want to fix the message to be? (It is not related to this PR)

> Expected message "Value transfer to a contract with a non-payable fallback function." but found "Value transfer to a contract without a payable fallback function."

**Comment by federicobond** [comment]
@chriseth rebased and updated.

**Comment by axic** [comment]
@chriseth are we good to merge this?

**Review by chriseth — APPROVED** [review]

**Pull request merged**